Content.Server/GameObjects/Components/Doors/ServerDoorComponent.cs

```diff
@@ -310,139 +356,147 @@ public void TryClose(IEntity user)
                 {
-                    airtight.AirBlocked = true;
+                    Open();
+                    return;
                 }
 
-                if (Owner.TryGetComponent(out IPhysicsComponent? body))
+                OnPartialClose();
+                await Timer.Delay(CloseTimeTwo, _cancellationTokenSource.Token);
+                if (Occludes && Owner.TryGetComponent(out OccluderComponent? occluder))
                 {
-                    body.CanCollide = true;
+                    occluder.Enabled = true;
                 }
 
-                await Timer.Delay(CloseTimeTwo, _cancellationTokenSource.Token);
-
-                _canWeldShut = true;
                 State = DoorState.Closed;
-                SetAppearance(DoorVisualState.Closed);
             }, _cancellationTokenSource.Token);
+        }
+
+        protected override void OnPartialClose()
+        {
+            base.OnPartialClose();
+
+            var becomeAirtight = true;
+            if (!Safety)
+            {
+                // crushes people inside of the door, temporarily turning off collisions with them while doing so, to avoid jank
+                becomeAirtight = !TryCrush();
+            }
+            if (becomeAirtight && Owner.TryGetComponent(out AirtightComponent? airtight))
+            {
+                airtight.AirBlocked = true;
+            }
+
             Owner.EntityManager.EventBus.RaiseEvent(EventSource.Local, new AccessReaderChangeMessage(Owner, true));
-            return true;
         }
 
-        public virtual void Deny()
+        /// <summary>
+        /// Crushes everyone colliding with us by more than 10%.
+        /// </summary>
+        /// <returns>True if we crushed somebody, false if we did not.</returns>
+        private bool TryCrush()
         {
-            if (State == DoorState.Open || _isWeldedShut)
+            if (!Owner.TryGetComponent(out IPhysicsComponent? physics))
+                return false;
+
+            // Crush
+            foreach (var e in physics.GetCollidingEntities(Vector2.Zero, false))
+            {
+                if (!e.TryGetComponent(out StunnableComponent? stun)
+                    || !e.TryGetComponent(out IDamageableComponent? damage)
+                    || !e.TryGetComponent(out IPhysicsComponent? otherBody))
+                    continue;
+
+                var percentage = otherBody.WorldAABB.IntersectPercentage(physics.WorldAABB);
+
+                if (percentage < 0.1f)
+                    continue;
+
+                CurrentlyCrushing = e.Uid;
+
+                damage.ChangeDamage(DamageType.Blunt, DoorCrushDamage, false, Owner);
+                stun.Paralyze(DoorStunTime);
+
+                // If we hit someone, open up after stun (opens right when stun ends)
+                Owner.SpawnTimer(TimeSpan.FromSeconds(DoorStunTime) - OpenTimeOne - OpenTimeTwo, Open);
+                return true;
+            }
+            return false;
+        }
```

Comment:
See elsewhere that it seems silly to only be able to crush one entity considering infinite entities can stand in a doorway and you're now relying on the physics system to nudge them out of the way.

You should also cache WorldAABB given it's expensive to calculate.